### PR TITLE
Boundary history class

### PR DESCRIPTION
**Pull request by wthrowe**
### Types of changes:

- [ ] Bugfix
- [X] New feature

### Component:

- [X] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

### CODE REVIEW

File: src/Time/BoundaryHistory.hpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  // because we also move the container being pointed into and maps
+  // guarantee that this doesn't invalidate pointers.
+  BoundaryHistory() = default;
+  BoundaryHistory(const BoundaryHistory&) = delete;
+  BoundaryHistory(BoundaryHistory&&) = default;
+  BoundaryHistory& operator=(const BoundaryHistory&) = delete;
+  BoundaryHistory& operator=(BoundaryHistory&&) = default;
+  ~BoundaryHistory() = default;
+
+  /// Modify history data on one of the sides.  See the corresponding
```
Comment (nilsdeppe) [inline src/Time/BoundaryHistory.hpp]: change comment

File: src/Time/BoundaryHistory.hpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+  // We use pointers instead of iterators because deque invalidates
+  // iterators when elements are inserted or removed at the ends, but
+  // not pointers.
+  std::map<std::pair<const Time*, const Time*>, CouplingResult> coupling_cache_;
+};
+
+template <typename LocalVars, typename RemoteVars, typename CouplingResult>
+void BoundaryHistory<LocalVars, RemoteVars, CouplingResult>::
+local_mark_unneeded(const local_iterator& first_needed) noexcept {
+  for (auto it = local_data_.begin(); it != first_needed.base(); ++it) {
```
Comment (nilsdeppe) [inline src/Time/BoundaryHistory.hpp]: Can the local and remote functions be combined?

File: tests/Unit/Time/Test_History.cpp [hunk 3]
```diff
@@ -153,3 +155,135 @@ SPECTRE_TEST_CASE("Unit.Time.History", "[Unit][Time]") {
...
+  history.remote_insert(make_time(2.), std::vector<int>{2});
+  history.remote_insert_initial(make_time(-2.), std::vector<int>{-2});
+  history.remote_insert(make_time(3.), std::vector<int>{3});
+
+  // Non-const arg for coupling caching
+  const auto check_state = [](HistoryType& hist) noexcept {
+    CHECK(hist.local_size() == 4);
+    {
+      auto it = hist.local_begin();
+      for (size_t i = 0; i < hist.local_size(); ++i) {
```
Comment (nilsdeppe) [inline tests/Unit/Time/Test_History.cpp]: ++i, ++it

File: tests/Unit/Time/Test_History.cpp [hunk 4]
```diff
@@ -153,3 +155,135 @@ SPECTRE_TEST_CASE("Unit.Time.History", "[Unit][Time]") {
...
+        CHECK(it->value() == entry_num);
+        ++it;
+      }
+      CHECK(it == hist.local_end());
+    }
+
+    CHECK(hist.remote_size() == 6);
+    {
+      auto it = hist.remote_begin();
+      for (size_t i = 0; i < hist.remote_size(); ++i) {
```
Comment (nilsdeppe) [inline tests/Unit/Time/Test_History.cpp]: ++i, ++it

File: tests/Unit/Time/Test_History.cpp [hunk 5]
```diff
@@ -153,3 +155,135 @@ SPECTRE_TEST_CASE("Unit.Time.History", "[Unit][Time]") {
...
+
+  history.remote_insert(make_time(1.), std::vector<int>{1});
+  history.remote_insert_initial(make_time(0.), std::vector<int>{0});
+  history.remote_insert_initial(make_time(-1.), std::vector<int>{-1});
+  history.remote_insert(make_time(2.), std::vector<int>{2});
+  history.remote_insert_initial(make_time(-2.), std::vector<int>{-2});
+  history.remote_insert(make_time(3.), std::vector<int>{3});
+
+  // Non-const arg for coupling caching
+  const auto check_state = [](HistoryType& hist) noexcept {
```
Comment (nilsdeppe) [inline tests/Unit/Time/Test_History.cpp]: turn into separate function

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 6]
```diff
@@ -319,238 +297,199 @@ void AdamsBashforthN::update_u(
...
+       ++local_evaluation_step) {
+    for (auto remote_evaluation_step = remote_start_step;
+         remote_evaluation_step != history->remote_end();
+         ++remote_evaluation_step) {
+      double deriv_coef = 0.;
+
+      // Prepare to loop over the union times relevant for the current
+      // coupling evaluation.
+
+      // Latest of start of current step (otherwise was computed in a
```
Comment (kidder) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: Latest time from either the time of the start of the current step (...

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 7]
```diff
@@ -319,238 +297,199 @@ void AdamsBashforthN::update_u(
...
+      // previous call) and the times of the side data being coupled
```
Comment (kidder) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: side -> remote (or remote side)

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 8]
```diff
@@ -319,238 +297,199 @@ void AdamsBashforthN::update_u(
...
+      // (since segments can't depend on later data).
+      const Time coupling_start_time =
+          std::max(start_time, *remote_evaluation_step, simulation_less);
+      // Earliest of the end of the current step (since the step can't
```
Comment (kidder) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: Earliest time from either the time at the end of the current ...

File: src/Time/TimeSteppers/AdamsBashforthN.hpp [hunk 9]
```diff
@@ -319,238 +297,199 @@ void AdamsBashforthN::update_u(
...
+      // depend on later data), and target_order steps after the
```
Comment (kidder) [inline src/Time/TimeSteppers/AdamsBashforthN.hpp]: and the times of (something about the remote data)

**Review by kidder — CHANGES_REQUESTED** [review]

**Review by kidder — COMMENTED** [review]
squash

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**